File: scripts/fetch_rain_history.py
#!/usr/bin/env python3
"""
Fetch historical rain data from Open-Meteo API for any plant.

Usage:
    python scripts/fetch_rain_history.py --plant-id eta
    python scripts/fetch_rain_history.py --plant-id ribera
    python scripts/fetch_rain_history.py --plant-id alpha1
"""

import argparse
import json
import requests
from datetime import datetime, timedelta
from pathlib import Path
import yaml


CLEANING_THRESHOLD_MM = 5.0
HEAVY_RAIN_THRESHOLD_MM = 10.0


def load_plant_config(plant_id: str) -> dict:
    """Load plant configuration from YAML file."""
    config_path = Path(f"plant_configs/{plant_id}.yaml")
    if not config_path.exists():
        raise FileNotFoundError(f"Plant config not found: {config_path}")

    with open(config_path) as f:
        return yaml.safe_load(f)


def fetch_rain_data(latitude: float, longitude: float, start_date: str, end_date: str) -> dict:
    """Fetch historical rain data from Open-Meteo Archive API."""
    url = "https://archive-api.open-meteo.com/v1/archive"

    params = {
        "latitude": latitude,
        "longitude": longitude,
        "start_date": start_date,
        "end_date": end_date,
        "daily": "precipitation_sum",
        "timezone": "auto"
    }

    print(f"Fetching rain data from Open-Meteo...")
    print(f"  Location: ({latitude}, {longitude})")
    print(f"  Period: {start_date} to {end_date}")

    response = requests.get(url, params=params)
    response.raise_for_status()

    return response.json()


def process_rain_data(api_data: dict, plant_id: str) -> dict:
    """Process API data into structured format."""
    daily = api_data.get("daily", {})
    dates = daily.get("time", [])
    precip = daily.get("precipitation_sum", [])

    if not dates or not precip:
        raise ValueError("No data returned from API")

    # Build daily data
    daily_data = []
    cleaning_events = []

    for date, rain_mm in zip(dates, precip):
        rain_mm = rain_mm if rain_mm is not None else 0.0

        is_cleaning = rain_mm >= CLEANING_THRESHOLD_MM
        is_heavy = rain_mm >= HEAVY_RAIN_THRESHOLD_MM

        daily_data.append({
            "date": date,
            "precipitation_mm": round(rain_mm, 1),
            "is_cleaning_event": is_cleaning,
            "is_heavy_rain": is_heavy
        })

        # Record cleaning events
        if is_cleaning:
            # Estimate SR recovery based on rain amount
            if is_heavy:
                expected_recovery = 0.15  # Heavy rain can recover ~15% SR
                event_type = "heavy"
            else:
                expected_recovery = 0.08  # Moderate rain ~8% SR recovery
                event_type = "moderate"

            cleaning_events.append({
                "date": date,
                "amount_mm": round(rain_mm, 1),
                "type": event_type,
                "expected_sr_recovery": expected_recovery
            })

    # Calculate statistics
    total_days = len(daily_data)
    rain_days = sum(1 for d in daily_data if d["precipitation_mm"] > 0.1)
    total_precip = sum(d["precipitation_mm"] for d in daily_data)
    cleaning_count = len(cleaning_events)

    # Build response
    return {
        "metadata": {
            "plant_id": plant_id,
            "latitude": api_data["latitude"],
            "longitude": api_data["longitude"],
            "period": {
                "start": dates[0],
                "end": dates[-1]
            },
            "source": "Open-Meteo Archive API",
            "cleaning_threshold_mm": CLEANING_THRESHOLD_MM,
            "generated_at": datetime.now().isoformat()
        },
        "daily_data": daily_data,
        "cleaning_events": cleaning_events,
        "statistics": {
            "total_days": total_days,
            "rain_days": rain_days,
            "rain_frequency": round(rain_days / total_days, 3) if total_days > 0 else 0,
            "cleaning_events_count": cleaning_count,
            "total_precipitation_mm": round(total_precip, 1),
            "avg_precipitation_mm": round(total_precip / total_days, 2) if total_days > 0 else 0
        }
    }


def main():
    parser = argparse.ArgumentParser(description="Fetch rain history from Open-Meteo")
    parser.add_argument("--plant-id", required=True, help="Plant ID")
    parser.add_argument("--start-date", help="Start date (YYYY-MM-DD), default: 3 years ago")
    parser.add_argument("--end-date", help="End date (YYYY-MM-DD), default: today")
    args = parser.parse_args()

    plant_id = args.plant_id

    print(f"\n{'='*60}")
    print(f"FETCHING RAIN HISTORY FOR {plant_id.upper()}")
    print(f"{'='*60}\n")

    # Load plant config
    config = load_plant_config(plant_id)
    latitude = config['location']['latitude']
    longitude = config['location']['longitude']

    # Default date range: 3 years ago to today
    end_date = args.end_date or datetime.now().strftime("%Y-%m-%d")
    if not args.start_date:
        start_date = (datetime.now() - timedelta(days=3*365)).strftime("%Y-%m-%d")
    else:
        start_date = args.start_date

    # Fetch data
    api_data = fetch_rain_data(latitude, longitude, start_date, end_date)

    # Process data
    rain_data = process_rain_data(api_data, plant_id)

    # Save output
    output_dir = Path(f"public/data/soiling/{plant_id}")
    output_dir.mkdir(parents=True, exist_ok=True)

    # Save JSON (full data with metadata)
    output_path_json = output_dir / "rain_history.json"
    with open(output_path_json, 'w') as f:
        json.dump(rain_data, f, indent=2)

    # Save CSV (simple format for per-inverter training)
    import pandas as pd
    output_path_csv = output_dir / "rain_history.csv"
    df_rain = pd.DataFrame(rain_data['daily_data'])
    df_rain.to_csv(output_path_csv, index=False)

    print(f"\n✅ Rain history saved to:")
    print(f"  JSON: {output_path_json}")
    print(f"  CSV:  {output_path_csv}")
    print(f"\nSummary:")
    print(f"  Total days: {rain_data['statistics']['total_days']}")
    print(f"  Rain days: {rain_data['statistics']['rain_days']} ({rain_data['statistics']['rain_frequency']*100:.1f}%)")
    print(f"  Cleaning events: {rain_data['statistics']['cleaning_events_count']}")
    print(f"  Total precipitation: {rain_data['statistics']['total_precipitation_mm']} mm")
    print(f"  Avg daily: {rain_data['statistics']['avg_precipitation_mm']} mm/day")


if __name__ == "__main__":
    main()

File: scripts/fetchRainHistory.ts
/**
 * Script to fetch historical rain data from Open-Meteo API
 * and save as JSON for the soiling intelligence feature.
 *
 * Usage: npx ts-node scripts/fetchRainHistory.ts
 */

const OPEN_METEO_BASE_URL = 'https://archive-api.open-meteo.com/v1/archive';
const CLEANING_THRESHOLD_MM = 5.0;
const HEAVY_RAIN_THRESHOLD_MM = 10.0;

interface RainDataPoint {
  date: string;
  precipitation_mm: number;
  is_cleaning_event: boolean;
  is_heavy_rain: boolean;
}

interface RainCleaningEvent {
  date: string;
  amount_mm: number;
  type: 'moderate' | 'heavy';
  expected_sr_recovery: number;
}

interface OpenMeteoResponse {
  latitude: number;
  longitude: number;
  daily: {
    time: string[];
    precipitation_sum: number[];
  };
}

interface RainHistoryData {
  metadata: {
    plant_id: string;
    latitude: number;
    longitude: number;
    period: { start: string; end: string };
    source: string;
    cleaning_threshold_mm: number;
    generated_at: string;
  };
  daily_data: RainDataPoint[];
  cleaning_events: RainCleaningEvent[];
  statistics: {
    total_days: number;
    rain_days: number;
    cleaning_events_count: number;
    heavy_rain_count: number;
    avg_monthly_precipitation: number;
  };
}

async function fetchRainData(
  lat: number,
  lon: number,
  startDate: string,
  endDate: string
): Promise<RainDataPoint[]> {
  const url = new URL(OPEN_METEO_BASE_URL);
  url.searchParams.set('latitude', lat.toString());
  url.searchParams.set('longitude', lon.toString());
  url.searchParams.set('start_date', startDate);
  url.searchParams.set('end_date', endDate);
  url.searchParams.set('daily', 'precipitation_sum');
  url.searchParams.set('timezone', 'auto');

  console.log(`Fetching rain data from ${startDate} to ${endDate}...`);

  const response = await fetch(url.toString());

  if (!response.ok) {
    throw new Error(`Open-Meteo API error: ${response.status} ${response.statusText}`);
  }

  const data: OpenMeteoResponse = await response.json();

  const rainData: RainDataPoint[] = data.daily.time.map((date, index) => {
    const precipitation = data.daily.precipitation_sum[index] ?? 0;
    return {
      date,
      precipitation_mm: parseFloat(precipitation.toFixed(1)),
      is_cleaning_event: precipitation >= CLEANING_THRESHOLD_MM,
      is_heavy_rain: precipitation >= HEAVY_RAIN_THRESHOLD_MM,
    };
  });

  return rainData;
}

function extractCleaningEvents(rainData: RainDataPoint[]): RainCleaningEvent[] {
  return rainData
    .filter(d => d.is_cleaning_event)
    .map(d => ({
      date: d.date,
      amount_mm: d.precipitation_mm,
      type: d.is_heavy_rain ? 'heavy' as const : 'moderate' as const,
      expected_sr_recovery: d.is_heavy_rain ? 0.05 : 0.02,
    }));
}

function calculateStatistics(rainData: RainDataPoint[]) {
  const totalDays = rainData.length;
  const rainDays = rainData.filter(d => d.precipitation_mm > 0).length;
  const cleaningEvents = rainData.filter(d => d.is_cleaning_event).length;
  const heavyRainDays = rainData.filter(d => d.is_heavy_rain).length;

  const totalPrecip = rainData.reduce((sum, d) => sum + d.precipitation_mm, 0);
  const months = totalDays / 30.44;
  const avgMonthlyPrecip = months > 0 ? totalPrecip / months : 0;

  return {
    total_days: totalDays,
    rain_days: rainDays,
    cleaning_events_count: cleaningEvents,
    heavy_rain_count: heavyRainDays,
    avg_monthly_precipitation: parseFloat(avgMonthlyPrecip.toFixed(1)),
  };
}

async function main() {
  // ALPHA1 plant configuration
  const plantId = 'alpha1';
  const latitude = 37.8145;
  const longitude = -3.8047;
  const startDate = '2019-01-01';
  const endDate = '2025-12-08';

  console.log(`\nFetching rain history for ${plantId}`);
  console.log(`Location: ${latitude}, ${longitude} (Andalusia, Spain)`);
  console.log(`Period: ${startDate} to ${endDate}\n`);

  try {
    // Fetch all rain data
    const dailyData = await fetchRainData(latitude, longitude, startDate, endDate);

    console.log(`Received ${dailyData.length} days of data`);

    // Extract cleaning events
    const cleaningEvents = extractCleaningEvents(dailyData);
    console.log(`Found ${cleaningEvents.length} natural cleaning events (>5mm)`);

    // Calculate statistics
    const statistics = calculateStatistics(dailyData);

    // Build final structure
    const rainHistory: RainHistoryData = {
      metadata: {
        plant_id: plantId,
        latitude,
        longitude,
        period: { start: startDate, end: endDate },
        source: 'Open-Meteo Historical Weather API',
        cleaning_threshold_mm: CLEANING_THRESHOLD_MM,
        generated_at: new Date().toISOString(),
      },
      daily_data: dailyData,
      cleaning_events: cleaningEvents,
      statistics,
    };

    // Save to file
    const fs = await import('fs');
    const path = await import('path');

    const outputDir = path.join(process.cwd(), 'public/data/soiling', plantId);
    const outputPath = path.join(outputDir, 'rain_history.json');

    // Ensure directory exists
    if (!fs.existsSync(outputDir)) {
      fs.mkdirSync(outputDir, { recursive: true });
    }

    fs.writeFileSync(outputPath, JSON.stringify(rainHistory, null, 2));

    console.log(`\nRain history saved to: ${outputPath}`);
    console.log('\nStatistics:');
    console.log(`  Total days: ${statistics.total_days}`);
    console.log(`  Rain days: ${statistics.rain_days}`);
    console.log(`  Cleaning events (>5mm): ${statistics.cleaning_events_count}`);
    console.log(`  Heavy rain days (>10mm): ${statistics.heavy_rain_count}`);
    console.log(`  Avg monthly precipitation: ${statistics.avg_monthly_precipitation}mm`);

    // Also create empty labels file
    const labelsPath = path.join(outputDir, 'labels.json');
    const emptyLabels = {
      plantId,
      labels: [],
      createdAt: new Date().toISOString(),
      lastModified: new Date().toISOString(),
    };
    fs.writeFileSync(labelsPath, JSON.stringify(emptyLabels, null, 2));
    console.log(`\nEmpty labels file created: ${labelsPath}`);

  } catch (error) {
    console.error('Error fetching rain data:', error);
    process.exit(1);
  }
}

main();
